plugin.js

```diff
@@ -9,59 +10,112 @@ function Plugin (parent, func, opts, callback) {
   this.deferred = false
   this.onFinish = null
   this.parent = parent
+  this.name = func.name
 
   this.q = fastq(parent, loadPlugin, 1)
   this.q.pause()
+  this.loaded = false
 
   // always start the queue in the next tick
   // because we try to attach subsequent call to use()
   // to the right plugin. we need to defer them,
   // or they will end up at the top of _current
-  process.nextTick(this.q.resume.bind(this.q))
 }
 
 Plugin.prototype.exec = function (server, cb) {
   const func = this.func
+  var completed = false
   this.server = this.parent.override(server, func, this.opts)
 
-  // we must defer the loading of the plugin until the
-  // current execution has ended
-  process.nextTick(() => {
-    func(this.server, this.opts, cb)
-  })
+  debug('exec', this.name)
+
+  var promise = func(this.server, this.opts, done)
+  if (promise && typeof promise.then === 'function') {
```

Comment:
We could add a debug log here, something like `debug('resolving promise', this.name)`.
Same few lines below: `debug('promise resolved', this.name)`.